coderd/database/queries/prebuilds.sql

```diff
@@ -0,0 +1,120 @@
+	 latest_builds AS (
+		 -- Select only the latest build per template_version AND preset
+		 SELECT fb.*,
+				ROW_NUMBER() OVER (PARTITION BY fb.template_version_preset_id ORDER BY fb.created_at DESC) as rn
+		 FROM filtered_builds fb),
+	 failed_count AS (
+		 -- Count failed builds per template version/preset in the given period
+		 SELECT preset_id, COUNT(*) AS num_failed
+		 FROM filtered_builds
+		 WHERE job_status = 'failed'::provisioner_job_status
+		   AND created_at >= @lookback::timestamptz
+		 GROUP BY preset_id)
+SELECT lb.template_version_id,
+	   lb.preset_id,
+	   MAX(lb.job_status)::provisioner_job_status AS latest_build_status,
+	   MAX(COALESCE(fc.num_failed, 0))::int       AS num_failed,
+	   MAX(lb.created_at)::timestamptz            AS last_build_at
+FROM latest_builds lb
+		 LEFT JOIN failed_count fc ON fc.preset_id = lb.preset_id
+WHERE lb.rn <= lb.desired_instances -- Fetch the last N builds, where N is the number of desired instances; if any fail, we backoff
+  AND lb.job_status = 'failed'::provisioner_job_status
+GROUP BY lb.template_version_id, lb.preset_id, lb.job_status;
+
+-- name: ClaimPrebuild :one
+UPDATE workspaces w
+SET owner_id   = @new_user_id::uuid,
+	name       = @new_name::text,
+	updated_at = NOW()
+WHERE w.id IN (SELECT p.id
+			   FROM workspace_prebuilds p
+						INNER JOIN workspace_latest_build b ON b.workspace_id = p.id
+						INNER JOIN provisioner_jobs pj ON b.job_id = pj.id
+						INNER JOIN templates t ON p.template_id = t.id
+			   WHERE (b.transition = 'start'::workspace_transition
+				   AND pj.job_status IN ('succeeded'::provisioner_job_status))
+				 AND b.template_version_id = t.active_version_id
+				 AND b.template_version_preset_id = @preset_id::uuid
+				 AND p.lifecycle_state = 'ready'::workspace_agent_lifecycle_state
+			   ORDER BY random()
+			   LIMIT 1 FOR UPDATE OF p SKIP LOCKED) -- Ensure that a concurrent request will not select the same prebuild.
+RETURNING w.id, w.name;
+
+-- name: InsertPresetPrebuild :one
+INSERT INTO template_version_preset_prebuilds (id, preset_id, desired_instances, invalidate_after_secs)
+VALUES (@id::uuid, @preset_id::uuid, @desired_instances::int, @invalidate_after_secs::int)
+RETURNING *;
+
+-- name: GetPrebuildMetrics :many
+SELECT
+    t.name as template_name,
+    tvp.name as preset_name,
+    COUNT(*) as created_count,
+    COUNT(*) FILTER (WHERE pj.job_status = 'failed'::provisioner_job_status) as failed_count,
+    COUNT(*) FILTER (
+			 WHERE w.owner_id != 'c42fdf75-3097-471c-8c33-fb52454d81c0'::uuid -- The system user responsible for prebuilds.
+		) as claimed_count
+FROM workspaces w
+INNER JOIN workspace_prebuild_builds wpb ON wpb.workspace_id = w.id
+INNER JOIN templates t ON t.id = w.template_id
+INNER JOIN template_version_presets tvp ON tvp.id = wpb.template_version_preset_id
+INNER JOIN provisioner_jobs pj ON pj.id = wpb.job_id
+WHERE wpb.build_number = 1
+GROUP BY t.name, tvp.name
```

Comment:
1) Template names are only unique within an organization.

 2) Templates are soft-deleted, so if a template is deleted and a new template created with the same name, this query will include the deleted one's stats.

Reply:
let's filter out deleted templates

Reply:
this is done, and we've added tests for it. the tests are part of a subsequent PR, because they are run at a higher level of abstraction.